### PR TITLE
fix(agent): fix enable user input bug

### LINKED ISSUE #626: agent: cannot disable user input for an agent
### Describe the bug

When we try to disable user input for an agent, we get the below error.

### Expected Behavior

User input can be disabled without any error.

### Current Behavior

We get the below error on CFn deployment.

> 13:45:52 | UPDATE_FAILED        | AWS::Bedrock::Agent | Agent1Agent01A464A5
Resource handler returned message: "Invalid request provided: ActionGroup with name UserInputAction cannot be deleted when it is Enabled. Either use SkipResourceInUseCheckOnDelete flag or mark it as Disabled." (RequestToken: , HandlerErrorCode: InvalidRequest)

### Reproduction Steps

Create an agent

```ts
import { bedrock } from '@cdklabs/generative-ai-cdk-constructs';

   new bedrock.Agent(this, 'Agent', {
      foundationModel: bedrock.BedrockFoundationModel.ANTHROPIC_CLAUDE_SONNET_V1_0,
      instruction: 'You are a helpful and friendly agent that answers questions about literature.',
      shouldPrepareAgent: true,
      enableUserInput: true,
    });
```

Run cdk deploy, and then change enableUserInput: false:

```diff
-      enableUserInput: true,
+      enableUserInput: false,
```

cdk deploy, and we get the above error.

### Possible Solution

As the error message suggests, we need to use SkipResourceInUseCheckOnDelete flag or mark it as Disabled. As the stateless nature of CDK, SkipResourceInUseCheckOnDelete seems to be a better option, but not sure the risk of defaulting it to true.

### Additional Information/Context

_No response_

### CDK CLI Version

2.151.0

### Framework Version

_No response_

### Node.js Version

v20.10.0

### OS

macOS

### Language

Typescript

### Language Version

_No response_

### Region experiencing the issue

us-west-2

### Code modification

no

### Other information

_No response_

### Service quota

- [X] I have reviewed the service quotas for this construct

### PR DESCRIPTION
**Pull request by aws-rafams**
Fixes #626 

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

### CODE REVIEW

**Comment by aws-rafams** [comment]
> I have understood what is the problem here, when `enableUserInput: true`, this action group is visible on the console under Agent builder -> Additional Settings -> User Input. This part gets created by this piece of code here:
> 
> https://github.com/awslabs/generative-ai-cdk-constructs/blob/6b5bbeb84de1515d620f88762276ad91c788b9c9/src/cdk-lib/bedrock/agent.ts#L492-L498
> 
> 
> the problem is that when the update happens, this content of this if statement is not executed and thus the ActionGroup is not synthesized, thus disappearing from the template and triggering DELETE operation, however, this delete operation is not allowed, because it should either be set to DISABLED (an update to the ActionGroup resource), or `skipResourceInUseCheckOnDelete` must be set to `true`. I believe the easiest solution to this bug is to set:
> ```diff
> +      skipResourceInUseCheckOnDelete: true,
> ```
> 
> during the ActionGroup Creation as in PR #642. Otherwise the other solution would be to always create the action group and according to whether it must be enabled, set the `actionGroupState` accordingly:
> 
> ```ts
> this.addActionGroup(new AgentActionGroup(this, 'userInputEnabledActionGroup', {
>     actionGroupName: 'UserInputAction',
>     parentActionGroupSignature: 'AMAZON.UserInput',
>     actionGroupState: props.enableUserInput ? 'ENABLED' :  'DISABLED',
> }));
> ```

**Comment by scottschreckengaust** [comment]
How might the second option operate? 

> ```
> actionGroupState: props.enableUserInput ? 'ENABLED' :  'DISABLED',
> ```

**Comment by aws-rafams** [comment]
> ```ts
> props.enableUserInput ? 'ENABLED' :  'DISABLED',
> ```

I have tested both solutions, and they are both functional. However, there is a subtle difference in their behaviour. With the second approach, the action group is always created (as demonstrated in the screenshot below), and the only variation lies in whether it is enabled or disabled.

Screenshot 2024-08-19 at 19 17 35

In any case this action group will not be shown in the console as an action group but rather as a configuration parameter on the agent. 

Screenshot 2024-08-19 at 19 18 48

**Comment by scottschreckengaust** [comment]
> > ```ts
> > props.enableUserInput ? 'ENABLED' :  'DISABLED',
> > ```
> 
> I have tested both solutions, and they are both functional. However, there is a subtle difference in their behaviour. With the second approach, the action group is always created (as demonstrated in the screenshot below), and the only variation lies in whether it is enabled or disabled.

Thank you very much. Upon review, would you mind updating for the second option?

**Comment by dineshSajwan** [comment]
@aws-rafams . Thank you so much for suggesting the fix. It looks good to me however option 2 by Scott also make sense. once you create the userinput on the console you can enable or disable it.  With option 1, Can we please verify that CDK destroy is removing all the resources with SkipResourceInUseCheckOnDelete set to true.

**Comment by scottschreckengaust** [comment]
Please review test results (specifically the `agent.test.ts`). Thank you.

```shell
% npx projen test --testPathPattern='.*agent\.test.*'
...
  Agent with guardrails through addGuardrail
    ✓ Knowledge Base is created (105 ms)
    ✓ Data Source is created (1 ms)
    ✓ Agent is created (1 ms)
    ✓ Agent is created with one knowledge base (1 ms)
    ✕ Agent action group and ApiSchema from S3 (2 ms)
    ✓ Guardrail is associated (17 ms)
    ✓ Agent Alias is created (1 ms)
    ✓ No unsuppressed Errors (58 ms)
  Agent with guardrails through constructor
    ✓ Knowledge Base is created (48 ms)
    ✓ Data Source is created (1 ms)
    ✓ Agent is created (1 ms)
    ✓ Agent is created with one knowledge base
    ✕ Agent action group and ApiSchema from S3 (1 ms)
    ✓ Guardrail is associated (3 ms)
    ✓ Agent Alias is created (1 ms)
    ✓ No unsuppressed Errors (46 ms)
  Agent without guardrails
    ✓ Knowledge Base is created (47 ms)
    ✓ Data Source is created (1 ms)
    ✓ Agent is created (1 ms)
    ✓ Agent is created with one knowledge base
    ✕ Agent action group and ApiSchema from S3 (1 ms)
    ✓ Guardrail should not be associated (1 ms)
    ✓ Agent Alias is created (1 ms)
    ✓ No unsuppressed Errors (44 ms)
...
```

File: src/cdk-lib/bedrock/agent.ts [hunk 1]
```diff
@@ -380,13 +380,13 @@ export class Agent extends Construct {
    * A list of action groups associated with the agent
    * @private
    */
-  public actionGroups: bedrock.CfnAgent.AgentActionGroupProperty[]=[];
+  public actionGroups: bedrock.CfnAgent.AgentActionGroupProperty[] = [];
```
Comment (scottschreckengaust) [inline src/cdk-lib/bedrock/agent.ts:383]: (non-blocking) Consider keeping formatting changes in a separate PR

Comment (aws-rafams) [reply src/cdk-lib/bedrock/agent.ts:383]: sorry, was not my intention, was formatted automatically on save by my code formatter

**Comment by aws-rafams** [comment]
tests have been updated

**Review by dineshSajwan — APPROVED** [review]
LGTM

**Comment by aws-rafams** [comment]
is anything missing for this PR to be merged?

**Comment by krokoko** [comment]
I will just test it today in my account and if everything is good will approve, thanks @aws-rafams !

**Review by krokoko — APPROVED** [review]
Tested by deploying an agent and following the steps described in the bug ticket, works as expected. Thanks @aws-rafams for your contribution!

**Pull request merged**